Branch-SDK/src/main/java/io/branch/referral/BranchPostTask.java

```diff
@@ -0,0 +1,194 @@
+                        branch.prefHelper_.setDeviceFingerPrintID(respJson.getString(Defines.Jsonkey.DeviceFingerprintID.getKey()));
+                        updateRequestsInQueue = true;
+                    }
+                    if (updateRequestsInQueue) {
+                        branch.updateAllRequestsInQueue();
+                    }
+                } catch (JSONException ex) {
+                    ex.printStackTrace();
+                }
+            }
+
+            if (thisReq_ instanceof ServerRequestInitSession) {
+                branch.setInitState(Branch.SESSION_STATE.INITIALISED);
+                if (!((ServerRequestInitSession) thisReq_).handleBranchViewIfAvailable((serverResponse))) {
+                    branch.checkForAutoDeepLinkConfiguration();
+                }
+                // Count down the latch holding getLatestReferringParamsSync
+                if (branch.getLatestReferringParamsLatch != null) {
+                    branch.getLatestReferringParamsLatch.countDown();
+                }
+                // Count down the latch holding getFirstReferringParamsSync
+                if (branch.getFirstReferringParamsLatch != null) {
+                    branch.getFirstReferringParamsLatch.countDown();
+                }
+            }
+        }
+
+        if (respJson != null) {
+            thisReq_.onRequestSucceeded(serverResponse, branch);
+            branch.requestQueue_.remove(thisReq_);
+        } else if (thisReq_.shouldRetryOnFail()) {
+            // already called handleFailure above
+            thisReq_.clearCallbacks();
+        } else {
+            branch.requestQueue_.remove(thisReq_);
+        }
+    }
+
+    private void onRequestFailed(ServerResponse serverResponse, int status) {
+        // If failed request is an initialisation request (but not in the intra-app linking scenario) then mark session as not initialised
+        if (thisReq_ instanceof ServerRequestInitSession && PrefHelper.NO_STRING_VALUE.equals(branch.prefHelper_.getSessionParams())) {
+            branch.setInitState(Branch.SESSION_STATE.UNINITIALISED);
+        }
+
+        // On a bad request or in case of a conflict notify with call back and remove the request.
+        if (status == 400 || status == 409 && thisReq_ instanceof ServerRequestCreateUrl) {
+            ((ServerRequestCreateUrl) thisReq_).handleDuplicateURLError();
+        } else {
+            //On Network error or Branch is down fail all the pending requests in the queue except
+            //for request which need to be replayed on failure.
+            branch.networkCount_ = 0;
+            thisReq_.handleFailure(status, serverResponse.getFailReason());
+        }
+
+        if (!thisReq_.shouldRetryOnFail()) {
+            branch.requestQueue_.remove(thisReq_);
+        } else {
+            // failure has already been handled
+            // todo does it make sense to retry the request without a callback? (e.g. CPID, LATD)
```

Comment:
@echo-branch the current implementation of CPID and LATD requests was such that they will retry on failure indefinitely but their clearCallbacks method was a no op, so potentially the user would just keep getting errors in the callback indefinitely. I fixed the `clearCallbacks` methods of those requests so they do what the method name says but wanted to double check with you and see what the iOS does.

Reply:
iOS doesn't retry them.

Reply:
Perfect!